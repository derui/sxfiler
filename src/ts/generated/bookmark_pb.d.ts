// package: 
// file: bookmark.proto

import * as jspb from "google-protobuf";

export class Bookmark extends jspb.Message {
  getName(): string;
  setName(value: string): void;

  getPath(): string;
  setPath(value: string): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): Bookmark.AsObject;
  static toObject(includeInstance: boolean, msg: Bookmark): Bookmark.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: Bookmark, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): Bookmark;
  static deserializeBinaryFromReader(message: Bookmark, reader: jspb.BinaryReader): Bookmark;
}

export namespace Bookmark {
  export type AsObject = {
    name: string,
    path: string,
  }
}

export class BookmarkList extends jspb.Message {
  clearItemsList(): void;
  getItemsList(): Array<Bookmark>;
  setItemsList(value: Array<Bookmark>): void;
  addItems(value?: Bookmark, index?: number): Bookmark;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): BookmarkList.AsObject;
  static toObject(includeInstance: boolean, msg: BookmarkList): BookmarkList.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: BookmarkList, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): BookmarkList;
  static deserializeBinaryFromReader(message: BookmarkList, reader: jspb.BinaryReader): BookmarkList;
}

export namespace BookmarkList {
  export type AsObject = {
    itemsList: Array<Bookmark.AsObject>,
  }
}

export class ListAllRequest extends jspb.Message {
  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): ListAllRequest.AsObject;
  static toObject(includeInstance: boolean, msg: ListAllRequest): ListAllRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: ListAllRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): ListAllRequest;
  static deserializeBinaryFromReader(message: ListAllRequest, reader: jspb.BinaryReader): ListAllRequest;
}

export namespace ListAllRequest {
  export type AsObject = {
  }
}

export class ListAllResponse extends jspb.Message {
  hasBookmarks(): boolean;
  clearBookmarks(): void;
  getBookmarks(): BookmarkList | undefined;
  setBookmarks(value?: BookmarkList): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): ListAllResponse.AsObject;
  static toObject(includeInstance: boolean, msg: ListAllResponse): ListAllResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: ListAllResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): ListAllResponse;
  static deserializeBinaryFromReader(message: ListAllResponse, reader: jspb.BinaryReader): ListAllResponse;
}

export namespace ListAllResponse {
  export type AsObject = {
    bookmarks?: BookmarkList.AsObject,
  }
}

export class RegisterRequest extends jspb.Message {
  getPath(): string;
  setPath(value: string): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): RegisterRequest.AsObject;
  static toObject(includeInstance: boolean, msg: RegisterRequest): RegisterRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: RegisterRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): RegisterRequest;
  static deserializeBinaryFromReader(message: RegisterRequest, reader: jspb.BinaryReader): RegisterRequest;
}

export namespace RegisterRequest {
  export type AsObject = {
    path: string,
  }
}

export class RegisterResponse extends jspb.Message {
  hasBookmark(): boolean;
  clearBookmark(): void;
  getBookmark(): Bookmark | undefined;
  setBookmark(value?: Bookmark): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): RegisterResponse.AsObject;
  static toObject(includeInstance: boolean, msg: RegisterResponse): RegisterResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: RegisterResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): RegisterResponse;
  static deserializeBinaryFromReader(message: RegisterResponse, reader: jspb.BinaryReader): RegisterResponse;
}

export namespace RegisterResponse {
  export type AsObject = {
    bookmark?: Bookmark.AsObject,
  }
}

export class DeleteRequest extends jspb.Message {
  getId(): string;
  setId(value: string): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): DeleteRequest.AsObject;
  static toObject(includeInstance: boolean, msg: DeleteRequest): DeleteRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: DeleteRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): DeleteRequest;
  static deserializeBinaryFromReader(message: DeleteRequest, reader: jspb.BinaryReader): DeleteRequest;
}

export namespace DeleteRequest {
  export type AsObject = {
    id: string,
  }
}

export class DeleteResponse extends jspb.Message {
  hasDeletedBookmarks(): boolean;
  clearDeletedBookmarks(): void;
  getDeletedBookmarks(): Bookmark | undefined;
  setDeletedBookmarks(value?: Bookmark): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): DeleteResponse.AsObject;
  static toObject(includeInstance: boolean, msg: DeleteResponse): DeleteResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: DeleteResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): DeleteResponse;
  static deserializeBinaryFromReader(message: DeleteResponse, reader: jspb.BinaryReader): DeleteResponse;
}

export namespace DeleteResponse {
  export type AsObject = {
    deletedBookmarks?: Bookmark.AsObject,
  }
}

