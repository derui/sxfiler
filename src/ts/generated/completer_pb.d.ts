// package: completer
// file: completer.proto

import * as jspb from "google-protobuf";

export class Item extends jspb.Message {
  getId(): string;
  setId(value: string): void;

  getValue(): string;
  setValue(value: string): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): Item.AsObject;
  static toObject(includeInstance: boolean, msg: Item): Item.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: Item, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): Item;
  static deserializeBinaryFromReader(message: Item, reader: jspb.BinaryReader): Item;
}

export namespace Item {
  export type AsObject = {
    id: string,
    value: string,
  }
}

export class Candidate extends jspb.Message {
  getStart(): number;
  setStart(value: number): void;

  getLength(): number;
  setLength(value: number): void;

  hasValue(): boolean;
  clearValue(): void;
  getValue(): Item | undefined;
  setValue(value?: Item): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): Candidate.AsObject;
  static toObject(includeInstance: boolean, msg: Candidate): Candidate.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: Candidate, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): Candidate;
  static deserializeBinaryFromReader(message: Candidate, reader: jspb.BinaryReader): Candidate;
}

export namespace Candidate {
  export type AsObject = {
    start: number,
    length: number,
    value?: Item.AsObject,
  }
}

export class InitializeRequest extends jspb.Message {
  clearSourceList(): void;
  getSourceList(): Array<Item>;
  setSourceList(value: Array<Item>): void;
  addSource(value?: Item, index?: number): Item;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): InitializeRequest.AsObject;
  static toObject(includeInstance: boolean, msg: InitializeRequest): InitializeRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: InitializeRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): InitializeRequest;
  static deserializeBinaryFromReader(message: InitializeRequest, reader: jspb.BinaryReader): InitializeRequest;
}

export namespace InitializeRequest {
  export type AsObject = {
    sourceList: Array<Item.AsObject>,
  }
}

export class InitializeResponse extends jspb.Message {
  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): InitializeResponse.AsObject;
  static toObject(includeInstance: boolean, msg: InitializeResponse): InitializeResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: InitializeResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): InitializeResponse;
  static deserializeBinaryFromReader(message: InitializeResponse, reader: jspb.BinaryReader): InitializeResponse;
}

export namespace InitializeResponse {
  export type AsObject = {
  }
}

export class CompleteRequest extends jspb.Message {
  getInput(): string;
  setInput(value: string): void;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): CompleteRequest.AsObject;
  static toObject(includeInstance: boolean, msg: CompleteRequest): CompleteRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: CompleteRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): CompleteRequest;
  static deserializeBinaryFromReader(message: CompleteRequest, reader: jspb.BinaryReader): CompleteRequest;
}

export namespace CompleteRequest {
  export type AsObject = {
    input: string,
  }
}

export class CompleteResponse extends jspb.Message {
  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): CompleteResponse.AsObject;
  static toObject(includeInstance: boolean, msg: CompleteResponse): CompleteResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: CompleteResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): CompleteResponse;
  static deserializeBinaryFromReader(message: CompleteResponse, reader: jspb.BinaryReader): CompleteResponse;
}

export namespace CompleteResponse {
  export type AsObject = {
  }
}

export class CompletionResultNotificationRequest extends jspb.Message {
  clearCandidatesList(): void;
  getCandidatesList(): Array<Candidate>;
  setCandidatesList(value: Array<Candidate>): void;
  addCandidates(value?: Candidate, index?: number): Candidate;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): CompletionResultNotificationRequest.AsObject;
  static toObject(includeInstance: boolean, msg: CompletionResultNotificationRequest): CompletionResultNotificationRequest.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: CompletionResultNotificationRequest, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): CompletionResultNotificationRequest;
  static deserializeBinaryFromReader(message: CompletionResultNotificationRequest, reader: jspb.BinaryReader): CompletionResultNotificationRequest;
}

export namespace CompletionResultNotificationRequest {
  export type AsObject = {
    candidatesList: Array<Candidate.AsObject>,
  }
}

export class CompletionResultNotificationResponse extends jspb.Message {
  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): CompletionResultNotificationResponse.AsObject;
  static toObject(includeInstance: boolean, msg: CompletionResultNotificationResponse): CompletionResultNotificationResponse.AsObject;
  static extensions: {[key: number]: jspb.ExtensionFieldInfo<jspb.Message>};
  static extensionsBinary: {[key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>};
  static serializeBinaryToWriter(message: CompletionResultNotificationResponse, writer: jspb.BinaryWriter): void;
  static deserializeBinary(bytes: Uint8Array): CompletionResultNotificationResponse;
  static deserializeBinaryFromReader(message: CompletionResultNotificationResponse, reader: jspb.BinaryReader): CompletionResultNotificationResponse;
}

export namespace CompletionResultNotificationResponse {
  export type AsObject = {
  }
}

