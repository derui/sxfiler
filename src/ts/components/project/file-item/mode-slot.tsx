import * as React from "react";
import { styled } from "@/components/theme";
import { CapabilityObject } from "@/domains/capability";
import { ModeObject } from "@/domains/mode";

function capabilityToString(cap: CapabilityObject) {
  const readable = cap.readable ? "r" : "-";
  const writable = cap.writable ? "w" : "-";
  const executable = cap.executable ? "x" : "-";

  return `${readable}${writable}${executable}`;
}

/**
 * convert mode to string
 */
function modeToString(mode: ModeObject, isDirectory: boolean, isSymlink: boolean): string {
  let state = "-";

  if (isDirectory && !isSymlink) {
    state = "d";
  } else if (isSymlink) {
    state = "l";
  }

  const owner = capabilityToString(mode.owner);
  const group = capabilityToString(mode.group);
  const others = capabilityToString(mode.others);

  return `${state}${owner}${group}${others}`;
}

interface Prop {
  mode: ModeObject;
  isDirectory: boolean;
  isSymlink: boolean;
}

const Mode = styled.span`
  flex: 0 1 auto;
  padding: 0 ${props => props.theme.spaces.small};

  white-space: nowrap;
`;

export const Component: React.FC<Prop> = prop => {
  const data = modeToString(prop.mode, prop.isDirectory, prop.isSymlink);

  return <Mode>{data}</Mode>;
};
