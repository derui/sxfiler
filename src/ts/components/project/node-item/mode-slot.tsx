import * as React from "react";
import { CapabilityObject } from "../../../domains/capability";
import { ModeObject } from "../../../domains/mode";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./node-item.module.scss");

function capabilityToString(cap: CapabilityObject) {
  const readable = cap.readable ? "r" : "-";
  const writable = cap.writable ? "w" : "-";
  const executable = cap.executable ? "x" : "-";

  return `${readable}${writable}${executable}`;
}

/**
 * convert mode to string
 */
function modeToString(mode: ModeObject, isDirectory: boolean, isSymlink: boolean): string {
  let state = " ";

  if (isDirectory && !isSymlink) {
    state = "d";
  } else if (isSymlink) {
    state = "l";
  }

  const owner = capabilityToString(mode.owner);
  const group = capabilityToString(mode.group);
  const others = capabilityToString(mode.others);

  return `${state}${owner}${group}${others}`;
}

interface Prop {
  mode: ModeObject;
  isDirectory: boolean;
  isSymlink: boolean;
}

const NodeMode: React.FC<Prop> = prop => {
  const data = modeToString(prop.mode, prop.isDirectory, prop.isSymlink);

  return <span className={styles.mode}>{data}</span>;
};

export default NodeMode;
