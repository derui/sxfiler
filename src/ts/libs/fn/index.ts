// define functional library in TypeScript

// pipe multiple function at once
export const pipe = function pipe<T extends any[], R>(fn1: (...args: T) => R, ...fns: Array<(arg: R) => R>) {
  const piped = fns.reduce(
    (current, next) => (value) => next(current(value)),
    (value) => value
  );

  return (...args: T) => piped(fn1(...args));
};

// compose two function
export const compose = function compose<V, R, N>(fn1: (arg: V) => R, fn2: (arg: R) => N) {
  return (v: V) => fn2(fn1(v));
};

/**
 * port Kotlin's `also` extend function.
 *
 * WARN: This function be able to modify the object in function. Avoid to use this funciton
 *       for shared function.
 */
export const also = function also<T>(it: T, f: (it: T) => void): T {
  f(it);
  return it;
};
