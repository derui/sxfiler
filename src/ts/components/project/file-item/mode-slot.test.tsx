import { h } from "preact";
import { render } from "preact-render-to-string";
import { Component as T } from "./mode-slot";
import { Mode, Capability } from "@/generated/filer_pb";
import { also } from "@/libs/fn";

describe("Project", () => {
  describe("Node Item", () => {
    describe("Mode slot", () => {
      it("should print name simply", () => {
        const mode = new Mode();
        const tree = render(<T mode={mode.toObject()} isDirectory={false} isSymlink={false} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print directory mode", () => {
        const mode = new Mode();
        mode.setOwner(
          also(new Capability(), (v) => {
            v.setReadable(true);
          })
        );
        const tree = render(<T mode={mode.toObject()} isDirectory={true} isSymlink={false} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print symlink mode", () => {
        const mode = new Mode();
        mode.setOwner(
          also(new Capability(), (v) => {
            v.setReadable(true);
          })
        );
        const tree = render(<T mode={mode.toObject()} isDirectory={false} isSymlink={true} />);

        expect(tree).toMatchSnapshot();
      });

      it("should print symlink when directory and symlink are true", () => {
        const mode = new Mode();
        mode.setOwner(
          also(new Capability(), (v) => {
            v.setReadable(true);
          })
        );

        const tree = render(<T mode={mode.toObject()} isDirectory={true} isSymlink={true} />);

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
