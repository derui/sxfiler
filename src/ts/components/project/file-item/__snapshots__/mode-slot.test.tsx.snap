// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node Item Mode slot should print directory mode 1`] = `"<span class=\\"file-item__item-mode\\" data-testid=\\"fileItem-modeSlot\\">dr--------</span>"`;

exports[`Project Node Item Mode slot should print name simply 1`] = `"<span class=\\"file-item__item-mode\\" data-testid=\\"fileItem-modeSlot\\">----------</span>"`;

exports[`Project Node Item Mode slot should print symlink mode 1`] = `"<span class=\\"file-item__item-mode\\" data-testid=\\"fileItem-modeSlot\\">lr--------</span>"`;

exports[`Project Node Item Mode slot should print symlink when directory and symlink are true 1`] = `"<span class=\\"file-item__item-mode\\" data-testid=\\"fileItem-modeSlot\\">lr--------</span>"`;
